apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/InitServiceNameInstrumentation2.java

```diff
@@ -0,0 +1,112 @@
+ */
+package co.elastic.apm.agent.servlet;
+
+import co.elastic.apm.agent.servlet.adapter.JakartaServletApiAdapter;
+import co.elastic.apm.agent.servlet.adapter.JavaxServletApiAdapter;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Set;
+
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.nameEndsWith;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.namedOneOf;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Instruments
+ * <ul>
+ *     <li>{@link javax.servlet.ServletContainerInitializer#onStartup(java.util.Set, javax.servlet.ServletContext)} </li>
+ *     <li>{@link jakarta.servlet.ServletContainerInitializer#onStartup(java.util.Set, jakarta.servlet.ServletContext)}</li>
+ * </ul>
+ * <p>
+ * Determines the service name based on the webapp's {@code META-INF/MANIFEST.MF} file early in the startup process.
+ * As this doesn't work with runtime attachment, the service name is also determined when the first request comes in.
+ */
+public abstract class InitServiceNameInstrumentation2 extends AbstractServletInstrumentation {
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return nameContains("Initializer");
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface()).and(hasSuperType(namedOneOf(
+            "javax.servlet.ServletContainerInitializer", "jakarta.servlet.ServletContainerInitializer")));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("onStartup")
+            .and(takesArguments(2))
+            .and(takesArgument(0, Set.class))
+            .and(takesArgument(1, nameEndsWith("ServletContext")));
+    }
+
+    public static class JavaxInitServiceNameInstrumentation extends InitServiceNameInstrumentation2 {
+
+        private static final JavaxServletApiAdapter adapter = JavaxServletApiAdapter.get();
+
+        @Override
+        public String rootClassNameThatClassloaderCanLoad() {
+            return "javax.servlet.AsyncContext";
+        }
+
+        public static class AdviceClass {
+
+            @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+            public static void onEnter(@Advice.Argument(1) @Nullable Object servletContext) {
```

Comment:
Instead of `Object`, you could use `javax.servlet.ServletContext`

Reply:
Using `javax.servlet.ServletContext`/`jakarta.servlet.ServletContext` as parameter will cause problems, if both are on the class path as both instrumentations will try to instrument the same class. But one of them will through an exception as the parameter can either be `javax.servlet.ServletContext` or `jakarta.servlet.ServletContext`.

Is having both on the class path something you will support?

Reply:
That's right, I didn't think about this. Let's leave it as-is then.